File: algorithms/python/213_house_robber_ii/house_robber_ii.py
"""
213. House Robber II

打家劫舍 II
"""

from typing import List


def rob(nums: List[int]) -> int:
    """
    动态规划
    因为首尾的特殊性，因此在 198 的前提下做两次动态规划，分成第一个房子偷与不偷的情况。
    如果偷第一个房子，那么最后一个房子一定不能偷，如果不偷第一个房子，那么不考虑第一个房子，剩下的正常考虑
    """

    def my_rob(nums: List[int]) -> int:
        """
        函数闭包
        """
        prev_max, cur_max = 0, 0
        for i in range(len(nums)):
            temp = cur_max
            cur_max = max(cur_max, prev_max + nums[i])
            prev_max = temp

        return cur_max

    return max(my_rob(nums[:-1]), my_rob(nums[1:])) if len(nums) != 1 else nums[0]

File: algorithms/213-house-robber-ii/house-robber-ii.go
package problem213

func max(x, y int) int {
	if x < y {
		return y
	}
	return x
}

// dynamicProgrammingRob 动态规划
func dynamicProgrammingRob(nums []int) int {
	prevMax := 0
	curMax := 0
	for _, num := range nums {
		temp := curMax
		curMax = max(curMax, prevMax+num)
		prevMax = temp
	}

	return curMax
}

func rob(nums []int) int {
	// 分两次动态规划
	// 1. 考虑偷第一间房子，那么最后一间房子一定不能偷
	// 2. 不考虑偷第一间房子
	length := len(nums)
	if length != 1 {
		return max(dynamicProgrammingRob(nums[:len(nums)-1]), dynamicProgrammingRob(nums[1:]))
	} else {
		return nums[0]
	}
}
